### PR TITLE
Issue 67: init command to generate Gitops directory and pipeline resources. 

### PR DESCRIPTION
**Pull request by chetan-rns**
**What type of PR is this?**

> /kind feature

**What does this PR do / why we need it**:

The new initialize command generates the agreed repository layout and creates the required pipeline resources.

`./odo pipelines init --gitops-repo org/repo --output ~/ --prefix foo --skip-checks`

The above command creates the following directory structure
carbon (4)

Changes:

 - Added `init` subcommand to `odo`
- Added functions to create the gitops directory layout
- Updated existing pipeline functions to create resources
- Added kustomize files which refer the pipeline resources

**Which issue(s) this PR fixes**:

Fixes https://issues.redhat.com/browse/GITOPS-67

Risks:
- This patch would break the existing `bootstrap` command

### CODE REVIEW

File: pkg/odo/cli/pipelines/initialise.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		Use:     name,
+		Short:   initialiseShortDesc,
+		Long:    initialiseLongDesc,
+		Example: fmt.Sprintf(initialiseExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			genericclioptions.GenericRun(o, cmd, args)
+		},
+	}
+
+	initialiseCmd.Flags().StringVar(&o.gitOpsRepo, "gitops-repo", "", "git repository in this form <username>/<repository>")
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/initialise.go]: "CI/CD pipelines configuration Git repository in this form \<username\>/\<repository\>"

Comment (chetan-rns) [reply pkg/odo/cli/pipelines/initialise.go]: I've created a separate `initialise.go` file for `init` command. Should I remove the` bootstrap.go` since it is not used anymore?

**Comment by chetan-rns** [comment]
> Please fix the failed tests.

Unit tests have been fixed.

File: pkg/odo/cli/pipelines/initialise.go (hunk 1 above)
Comment (wtam2018) [reply pkg/odo/cli/pipelines/initialise.go]: Yes.

File: pkg/odo/cli/cli.go [hunk 2]
```diff
@@ -174,7 +174,7 @@ func NewCmdOdo(name, fullName string) *cobra.Command {
 		preference.NewCmdPreference(preference.RecommendedCommandName, util.GetFullName(fullName, preference.RecommendedCommandName)),
 		debug.NewCmdDebug(debug.RecommendedCommandName, util.GetFullName(fullName, debug.RecommendedCommandName)),
 		pipelines.NewCmdComponent(pipelines.RecommendedCommandName, util.GetFullName(fullName, pipelines.RecommendedCommandName)),
-		pipelines.NewCmdBootstrap(pipelines.BootstrapRecommendedCommandName, util.GetFullName(fullName, pipelines.BootstrapRecommendedCommandName)),
+		pipelines.NewCmdInitialise(pipelines.InitialiseRecommendedCommandName, util.GetFullName(fullName, pipelines.InitialiseRecommendedCommandName)),
```
Comment (bigkevmcd) [inline pkg/odo/cli/cli.go]: I think we should rename this to `init`.

And the files to `init` too, instead of `initialise`.

File: pkg/odo/cli/pipelines/initialise.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	initialiseExample = ktemplates.Examples(`
+	# Initialise OpenShift pipelines in a cluster
+	%[1]s 
+	`)
+
+	initialiseLongDesc  = ktemplates.LongDesc(`Initialise GitOps CI/CD Pipelines`)
+	initialiseShortDesc = `Initialise pipelines`
+)
+
+// InitialiseParameters encapsulates the paratmeters for the odo pipelines initialise
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/initialise.go]:
```suggestion
// InitialiseParameters encapsulates the parameters for the odo pipelines initialise
```

File: pkg/odo/cli/pipelines/initialise.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// command.
+type InitialiseParameters struct {
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/initialise.go]: We could add comments to cover what some of these are used for.

File: pkg/pipelines/initialise_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+func TestWriteResources(t *testing.T) {
+	tmpDir, cleanUp := makeTempDir(t)
+	defer cleanUp()
+	resources := map[string][]interface{}{
+		"01_roles/serviceaccount.yaml": []interface{}{fakeYamlDoc(1)},
+		"02_tasks/buildah_task.yaml":   []interface{}{fakeYamlDoc(1), fakeYamlDoc(2)},
+	}
+
+	_, err := writeResources(tmpDir, "", resources)
```
Comment (bigkevmcd) [inline pkg/pipelines/initialise_test.go]: There's no test for the second parameter here.

But see above, I don't think we even need the second parameter here.

File: pkg/pipelines/initialise.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+func getCICDDir(path, prefix string) string {
+	return filepath.Join(path, envsDir, addPrefix(prefix, cicdDir))
+}
+
+func createPipelineResources(namespaces map[string]string, gitopsRepo, prefix string) map[string][]interface{} {
+
+	// key: path of the resource
+	// value: YAML content of the resource
+	outputs := make(map[string][]interface{}, 0)
```
Comment (bigkevmcd) [inline pkg/pipelines/initialise.go]: You can shorten this to

```suggestion
	outputs := map[string][]interface{}{}
```

File: pkg/pipelines/initialise.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	eventListener := eventlisteners.Generate(gitopsRepo, namespaces["cicd"], saName)
+	outputs[eventListenerPath] = append(outputs[eventListenerPath], eventListener)
+
+	route := routes.Generate(namespaces["cicd"])
+	outputs[routePath] = append(outputs[routePath], route)
+
+	return outputs
+}
+
+func writeResources(path string, prefix string, files map[string][]interface{}) ([]string, error) {
```
Comment (bigkevmcd) [inline pkg/pipelines/initialise.go]: I don't think you actually need to add the prefix to all the files.

The resources need to be prefixed, but not necessarily the filenames.

**Comment by bigkevmcd** [comment]
The biggest problem with this branch is that it's merging 67 and 68 in one branch, so it's not really doing the story.

We will likely want a separate "clean" repo output, so that we can use that to drive issue 74.

File: pkg/pipelines/init.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+	if err := addKustomize("bases", []string{"./pipelines"}, filepath.Join(getCICDDir(gitopsPath, o.Prefix), kustomize), o.Prefix); err != nil {
+		return err
+	}
+
+	if err := addKustomize("bases", []string{}, filepath.Join(gitopsPath, envsDir, baseDir, kustomize), o.Prefix); err != nil {
+		return err
+	}
+
+	fmt.Println("\nGitops initialised")
```
Comment (wtam2018) [inline pkg/pipelines/init.go]: I think we can get rid of this to make it less verbose.

Comment (chetan-rns) [reply pkg/pipelines/init.go]: Removed the print statement
